### ISSUE TITLE
Replace Dictionary<,>.Keys.Contains with ContainsKey

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/roslyn-analyzers#3467]
This is a proposal for a new analyzer/codefix for:

```c#
// given
Dictionary<int, string> dic = ...;

// this is O(N)
dic.Keys.Contains(1);

// this is O(1)
dic.ContainsKey(1);
```

I found this code in some design-time build targets today.

**Comment by mavasani** [comment]
@jeffhandley @terrajobst want to move this proposal to runtime repo under your radar?

**Comment by Evangelink** [comment]
I can take care of this implementation if you wish. I am wondering if we could somehow improve the banned API analyzer to allow catching such cases, basically you would want to forbid a specific method call in some specific context...

**Comment by drewnoakes** [comment]
> basically you would want to forbid a specific method call in some specific context...

In this case I think it would be enough to set the default diagnostic level to warning.

**Comment by Evangelink** [comment]
@drewnoakes Do you think there are cases where it would make sense to call `Contains` on a `Dictionary<TKey,TValue>.KeyCollection` or `Dictionary<TKey,TValue>.ValueCollection`? If not maybe the easiest is to actually ban `System.Collections.Generic.Dictionary<TKey,TValue>.KeyCollection.Contains(TKey)` and `System.Collections.Generic.Dictionary<TKey,TValue>.ValueCollection.Contains(TValue)`. WDYT?

Tagging @mavasani for a second opinion.

**Comment by drewnoakes** [comment]
The banned API analyzer tends to be for banning APIs that are specific to a given code base. I don't think any code base has a use case for calling either of those methods. It may be possible to ship a framework-specific analyzer that's based on the banned API analyzer and bundles a known set of APIs, such as these.

**Comment by Evangelink** [comment]
@drewnoakes My idea behind the use of the "Banned API" analyzer was to provide a way where we won't have to implement a new analyzer for each new member call we want to prevent. At the same time, the code in itself is pretty easy to write so there might be no point in having an infrastructure in place.

**Comment by Evangelink** [comment]
@drewnoakes @mavasani Do you think that's better to report only on specific dictionary instances or also on the interface? Reporting also on the interface might catch more cases but could potentially lead to FP for custom implementations of the dictionary. I think the risk is pretty low so I would be in favor of catching everything that implements `IDictionary<TKey, TValue>`/`IReadOnlyDictionary<TKey, TValue>`. WDYT?

**Comment by Evangelink** [comment]
In addition to my previous comment, I am wondering if we should extend the rule to also cover `Count()`. For `IDictionary` and `Dictionary` that seems to be fine because `Keys` returns `KeyCollection` with the property so there is another rule kicking-in to say use the property instead. But for `Keys` where the type is `IEnumerable` there is no rule reporting.

**Comment by mavasani** [comment]
@Evangelink This issue likely needs to be moved to https://github.com/dotnet/runtime with [code-analyzer](https://github.com/dotnet/runtime/labels/code-analyzer) label and would need an approval on the design and/or implementation.

@jeffhandley @bartonjs @stephentoub Can one of you please port this issue OR give an approval on the proposal here?
